crypto/s2n_certificate.c

```diff
@@ -35,6 +36,208 @@ int s2n_cert_set_cert_type(struct s2n_cert *cert, s2n_cert_type cert_type)
+    int rc = s2n_cert_chain_and_key_set_cert_chain_from_stuffer(cert_and_key, &chain_in_stuffer);
+
+    GUARD(s2n_stuffer_free(&chain_in_stuffer));
+
+    return rc;
+}
+
+int s2n_cert_chain_and_key_set_private_key(struct s2n_cert_chain_and_key *cert_and_key, const char *private_key_pem)
+{
+    DEFER_CLEANUP(struct s2n_stuffer key_in_stuffer = {{0}}, s2n_stuffer_free);
+    DEFER_CLEANUP(struct s2n_stuffer key_out_stuffer = {{0}}, s2n_stuffer_free);
+    struct s2n_blob key_blob = {0};
+
+    GUARD(s2n_pkey_zero_init(cert_and_key->private_key));
+
+    /* Put the private key pem in a stuffer */
+    GUARD(s2n_stuffer_alloc_ro_from_string(&key_in_stuffer, private_key_pem));
+    GUARD(s2n_stuffer_growable_alloc(&key_out_stuffer, strlen(private_key_pem)));
+
+    /* Convert pem to asn1 and asn1 to the private key. Handles both PKCS#1 and PKCS#8 formats */
+    GUARD(s2n_stuffer_private_key_from_pem(&key_in_stuffer, &key_out_stuffer));
+    key_blob.size = s2n_stuffer_data_available(&key_out_stuffer);
+    key_blob.data = s2n_stuffer_raw_read(&key_out_stuffer, key_blob.size);
+    notnull_check(key_blob.data);
+
+    /* Get key type and create appropriate key context */
+    GUARD(s2n_asn1der_to_private_key(cert_and_key->private_key, &key_blob));
+
+    return 0;
+}
+
+int s2n_cert_chain_and_key_set_ocsp_data(struct s2n_cert_chain_and_key *chain_and_key, const uint8_t *data, uint32_t length)
+{
+    notnull_check(chain_and_key);
+    GUARD(s2n_free(&chain_and_key->ocsp_status));
+    if (data && length) {
+        GUARD(s2n_alloc(&chain_and_key->ocsp_status, length));
+        memcpy_check(chain_and_key->ocsp_status.data, data, length);
+    }
+    return 0;
+}
+
+int s2n_cert_chain_and_key_set_sct_list(struct s2n_cert_chain_and_key *chain_and_key, const uint8_t *data, uint32_t length)
+{
+    notnull_check(chain_and_key);
+    GUARD(s2n_free(&chain_and_key->sct_list));
+    if (data && length) {
+        GUARD(s2n_alloc(&chain_and_key->sct_list, length));
+        memcpy_check(chain_and_key->sct_list.data, data, length);
+    }
+    return 0;
+}
+
+struct s2n_cert_chain_and_key *s2n_cert_chain_and_key_new(void)
+{
+    struct s2n_cert_chain_and_key *chain_and_key;
+    struct s2n_blob mem;
+    GUARD_PTR(s2n_alloc(&mem, sizeof(struct s2n_cert_chain_and_key)));
+    chain_and_key = (struct s2n_cert_chain_and_key *)(void *)mem.data;
+    
+    chain_and_key->cert_chain = NULL;
+    chain_and_key->private_key = NULL;
+    memset(&chain_and_key->ocsp_status, 0, sizeof(chain_and_key->ocsp_status));
+    memset(&chain_and_key->sct_list, 0, sizeof(chain_and_key->sct_list));
+    
+    return chain_and_key;
+}
+
+int s2n_cert_chain_and_key_init(struct s2n_cert_chain_and_key *chain_and_key, const char *chain_pem, const char *private_key_pem)
+{
+    struct s2n_blob cert_chain_mem, pkey_mem;
+
+    /* Allocate the memory for the chain and key */
+    GUARD(s2n_alloc(&cert_chain_mem, sizeof(struct s2n_cert_chain)));
```

Comment:
nit: I think it makes more sense to do this allocation in `s2n_cert_chain_and_key_new`. It could open the door for an `*_wipe` api for this struct in the future